core/src/test/scala/unit/kafka/server/GroupCoordinatorBaseRequestTest.scala

```diff
@@ -220,31 +292,92 @@ class GroupCoordinatorBaseRequestTest(cluster: ClusterInstance) {
     joinGroupResponse = connectAndReceive[JoinGroupResponse](joinGroupRequest)
     assertEquals(Errors.NONE.code, joinGroupResponse.data.errorCode)
 
-    val syncGroupRequestData = new SyncGroupRequestData()
-      .setGroupId(groupId)
-      .setMemberId(joinGroupResponse.data.memberId)
-      .setGenerationId(joinGroupResponse.data.generationId)
-      .setProtocolType("consumer")
-      .setProtocolName("consumer-range")
-      .setAssignments(List.empty.asJava)
-
-    // Send the sync group request to complete the rebalance.
-    val syncGroupRequest = new SyncGroupRequest.Builder(syncGroupRequestData).build()
-    val syncGroupResponse = connectAndReceive[SyncGroupResponse](syncGroupRequest)
-    assertEquals(Errors.NONE.code, syncGroupResponse.data.errorCode)
+    if (completeRebalance) {
+      // Send the sync group request to complete the rebalance.
+      syncGroupWithOldProtocol(
+        groupId = groupId,
+        memberId = joinGroupResponse.data.memberId(),
+        generationId = joinGroupResponse.data.generationId()
+      )
+    }
 
     (joinGroupResponse.data.memberId, joinGroupResponse.data.generationId)
   }
 
   protected def joinConsumerGroupWithNewProtocol(groupId: String): (String, Int) = {
-    // Heartbeat request to join the group.
+    val consumerGroupHeartbeatResponseData = consumerGroupHeartbeat(
+      groupId = groupId,
+      rebalanceTimeoutMs = 5 * 60 * 1000,
+      subscribedTopicNames = List("foo"),
+      topicPartitions = List.empty
+    )
+    (consumerGroupHeartbeatResponseData.memberId, consumerGroupHeartbeatResponseData.memberEpoch)
+  }
+
+  protected def joinConsumerGroup(groupId: String, useNewProtocol: Boolean): (String, Int) = {
+    if (useNewProtocol) {
+      // Note that we heartbeat only once to join the group and assume
+      // that the test will complete within the session timeout.
+      joinConsumerGroupWithNewProtocol(groupId)
+    } else {
+      // Note that we don't heartbeat and assume that the test will
+      // complete within the session timeout.
+      joinConsumerGroupWithOldProtocol(groupId)
+    }
+  }
+
+  protected def listGroups(
+    statesFilter: List[String],
+    version: Short
+  ): List[ListGroupsResponseData.ListedGroup] = {
+    val request = new ListGroupsRequest.Builder(
+      new ListGroupsRequestData()
+        .setStatesFilter(statesFilter.asJava)
+    ).build(version)
+
+    val response = connectAndReceive[ListGroupsResponse](request)
+
+    response.data().groups.asScala.toList
+  }
+
+  protected def describeGroups(
+    groupIds: List[String],
+    version: Short
+  ): List[DescribeGroupsResponseData.DescribedGroup] = {
+    val describeGroupsRequest = new DescribeGroupsRequest.Builder(
+      new DescribeGroupsRequestData()
+        .setGroups(groupIds.asJava)
+    ).build(version)
+
+    val describeGroupsResponse = connectAndReceive[DescribeGroupsResponse](describeGroupsRequest)
+
+    describeGroupsResponse.data().groups().asScala.toList
```

Comment:
nit: In Scala, we don't have to put `()` for getters so we usually don't put them in our code base. Here you can use `describeGroupsResponse.data.groups.asScala.toList`.

There are other cases where you could remove them. Could you please look at them?